### PR TITLE
Minimal codegen support: numbers and function parameters

### PR DESCRIPTION
**Pull request by Wybxc**
This is a minimal yet complete C code generation implementation, capable of compiling functions with number-typed parameters and `as` expressions.

The structure is organized as follows:
- **`crates/rustc_codegen_c`**: Implementation of the codegen backend.
- **`crates/rustc_codegen_c_ast`**: C syntax structure for generating C source code.
- **`bootstrap`**: Build system for executing `rustc` with this codegen backend enabled.
- **`example`**: Example files demonstrating the capabilities of the codegen backend.
- **`tests`**: Test cases for validating functionality.

### CODE REVIEW

**Comment by tgross35** [comment]
Thanks for getting this started! Is it possible to put bootstrap and CI into its own PR so that can be merged first, before the codegen part?

(it will take me a little while to review this all, still don't have my usual github access)

**Comment by Wybxc** [comment]
Understood, I'm working on #3

**Comment by tgross35** [comment]
This can be rebased now that #3 merged. If it is possible to split this further for smaller review units then that would be great (e.g. AST crate first and codegen separately), but if not than this shouldn't be too bad with bootstrap out of the picture.

**Comment by tgross35** [comment]
Could you put the license files into a separate PR that we can merge right away? Guess we are currently unlicensed.

Otherwise I'll try to look at this tomorrow-ish.

File: README.md
Comment (tgross35) [inline README.md]: Slight wording

```rust
This a C codegen backend for rustc, which lowers Rust MIR to C code and compiles
it with a C compiler.
```

File: .rustfmt.toml
Comment (tgross35) [inline .rustfmt.toml]: This file and the formatting updates to already-merged code should go in their own PR, to separate formatting changes from additions.

File: bootstrap/src/test.rs
Comment (tgross35) [inline bootstrap/src/test.rs]: Could you add doc comments for the variants?

File: bootstrap/src/test.rs
Comment (tgross35) [inline bootstrap/src/test.rs]: `let output_dir = self.output.parent().unwrap()` since it is used twice. But why do you need to navigate to the parent directory for this?

File: example/example.rs
Comment (tgross35) [inline example/example.rs]: Maybe name this file `basic_math.rs` or something, rather than just `example` since it's already in that directory.

Also `example` -> `examples` for the directory name, since there will be more than one :)

File: example/mini_core.rs
Comment (tgross35) [inline example/mini_core.rs]: This file has a lot of extra comments that should be removed before merge. Also should get a `//!` comment saying what its purpose is, even though it's semi-obvious.

**Review by tgross35 — COMMENTED** [review]
Added some initial comments, but I still have a lot more to go through.

Could you please add documentation comments? Most everything that isn't a trait implementation or a trivial function could benefit (it's a bit tough to figure out how things fit together without them).

File: crates/rustc_codegen_c_ast/src/arena.rs
Comment (tgross35) [inline crates/rustc_codegen_c_ast/src/arena.rs]: The module can just be renamed `ty` or `kind` or `type_` to avoid raw identifiers.

File: crates/rustc_codegen_c_ast/src/expr.rs
Comment (tgross35) [inline crates/rustc_codegen_c_ast/src/expr.rs]: Why is a `Scalar` different than a `Local`? Some docs or examples would help here.

File: crates/rustc_codegen_c_ast/src/lib.rs
Comment (tgross35) [inline crates/rustc_codegen_c_ast/src/lib.rs]: This module needs some docs, especially to clarify that the goal is just to be able to print (rather than parse).

File: crates/rustc_codegen_c_ast/src/lib.rs
Comment (tgross35) [inline crates/rustc_codegen_c_ast/src/lib.rs]: Nit: `ctx` is usually the abbreviation we use in the compiler (rather than `ctxt`)

File: crates/rustc_codegen_c_ast/src/decl.rs
Comment (tgross35) [inline crates/rustc_codegen_c_ast/src/decl.rs]: Any sort of type like this that has a C representation should get a small example of produced code in the doc comment (i.e. for each enum variant here).

File: crates/rustc_codegen_c_ast/src/decl.rs
Comment (tgross35) [inline crates/rustc_codegen_c_ast/src/decl.rs]: I think this logic should be reversed. That is, currently the whole thing looks like:

```rust
impl Printer {
    fn print_decl(&mut self, ...) { /* ... */ }
    fn print_other_thing(&mut self, ...) { /* ... */ }
    // ...
}
```

But this should be moved to a trait:

```rust
// Renamed from `Printer`
struct PrintCtx { pp: /* ... */ }

trait Print {
    fn print(&self, ptx: &mut PrintCtx);
}
```

And then that gets implemented for `CDecl` and all the other types.

File: crates/rustc_codegen_c_ast/src/pretty.rs
Comment (tgross35) [inline crates/rustc_codegen_c_ast/src/pretty.rs]: In the interest of time it is okay to not do the most fine-grained testing, but something is needed here. I would suggest you just add some simple output tests; in each module you can have a couple small tests that just construct an AST and print it to a string, then compare against a saved file. If `RUSTC_BLESS` is set, overwrite the file.

**Comment by tgross35** [comment]
`rustc_codegen_c_ast` is pretty self-contained, it would be good to move that to its own PR as well.

File: tests/codegen/filename.rs
Comment (tgross35) [inline tests/codegen/filename.rs:6]: In addition to the filecheck tests, you should add a couple of blessed tests. Add `tests/codegen-output/`, teach bootstrap to codegen all the `.rs` files and assert their output matches a `.c` of the same name in the same directory. And then add a `--bless` argument to bootstrap that will update the `.c` file.

Don't use that for complex tests where the MIR output is likely to change. But this would be good for simple tests like this one, and also is nice to show what the expected output is.

File: tests/codegen/ret_value.rs
Comment (tgross35) [inline tests/codegen/ret_value.rs]:
```rust
// CHECK-LABEL: main
// CHECK: return 42;
#[no_mangle]
pub fn main() -> i32 {
```

It is usually always good to use ` CHECK-LABEL`. And I assume you want to check that it actually returns 42, not just that it's somewhere in the code :)

File: tests/codegen/filename.rs
Comment (tgross35) [reply tests/codegen/filename.rs:6]: Also - these tests should get a comment about what they are intended to test. I'm not sure what this one is supposed to do.

File: tests/codegen/params_count.rs
Comment (tgross35) [inline tests/codegen/params_count.rs]: You probably want to check the full signature? Something like

```rust
// CHECK: int32_t foo(int32_t {{\w+}}, int32_t {{\w+}}, int32_t {{\w+}}) {
// CHECK-NEXT: return 0;
#[no_mangle]
pub fn foo(_x: i32, _y: i32, _z: i32) -> i32 {
```

Since filecheck only checks for partial matches .

File: example/mini_core.rs
Comment (tgross35) [reply example/mini_core.rs]: `tests/auxiliary/` would probably be a better place for this file, since it's not really an example. And this would mirror https://github.com/rust-lang/rust/pull/130693.

File: crates/rustc_codegen_c/src/util.rs
Comment (tgross35) [inline crates/rustc_codegen_c/src/util.rs:1]: This file gets emptied, any reason to keep it?

File: crates/rustc_codegen_c/src/helper.h
Comment (tgross35) [inline crates/rustc_codegen_c/src/helper.h]: Could you add a file-level doc comment explaining what `helper.h` is?

File: .rustfmt.toml
Comment (Wybxc) [reply .rustfmt.toml]: Ok, I'm making #5

File: bootstrap/src/test.rs
Comment (tgross35) [reply bootstrap/src/test.rs]: @Wybxc could you clarify why this needs to change to the parent directory of `output`? That doesn't seem to make sense.

**Comment by tgross35** [comment]
https://github.com/rust-lang/rustc_codegen_c/pull/6 merged so this can be rebased again.

It may be nice to just close this and open a new PR from the same branch to start fresh, since many of the comments here are related to things that have already merged.

File: README.md
Comment (tgross35) [inline README.md]: I would just say "This code is still highly experimental" or something like that, rather than saying some code isn't implemented. Since hopefully that will change :)

File: bootstrap/src/test.rs
Comment (tgross35) [inline bootstrap/src/test.rs]: Are aux crates just compiled for everything?

In rust-lang/rust we use a pattern like `//@ aux-build:mini_core.rs` ([example](https://github.com/rust-lang/rust/blob/b13176595d8a9fd27334747080764a53a18f4c24/tests/codegen/nounwind.rs#L1)). That would probably be worthwhile here to have better control over what is built / linked.

Just search the test files with regex `^//@\s*aux-build:(?P<fname>.*)`.

File: bootstrap/src/test.rs
Comment (tgross35) [inline bootstrap/src/test.rs]: Could this use something like [`similar`](https://docs.rs/similar/2.6.0/similar/) for better output?

File: crates/rustc_codegen_c/src/context.rs
Comment (tgross35) [inline crates/rustc_codegen_c/src/context.rs:39]: Could you add a doc comment about this item?

File: crates/rustc_codegen_c/src/helper.h
Comment (tgross35) [inline crates/rustc_codegen_c/src/helper.h:12]: Could you expand the docs with a bit about what this is doing? (I know you have it in the rust doc comments, but it would be good to have it next to the source).

File: crates/rustc_codegen_c/src/lib.rs
Comment (tgross35) [inline crates/rustc_codegen_c/src/lib.rs]: You can probably just drop this comment, I don't think there is any point in worrying about parallel codegen at this point.

File: crates/rustc_codegen_c/src/write.rs
Comment (tgross35) [inline crates/rustc_codegen_c/src/write.rs]: Delete unused comment

File: crates/rustc_codegen_c/src/write.rs
Comment (tgross35) [inline crates/rustc_codegen_c/src/write.rs]: Change `TODO` to `FIXME` if these are intended to be left in the codebase.

File: crates/rustc_codegen_c/src/write.rs
Comment (tgross35) [inline crates/rustc_codegen_c/src/write.rs]: Remove unused file.

File: tests/auxiliary/mini_core.rs
Comment (tgross35) [inline tests/auxiliary/mini_core.rs]: Delete these unused comments before merge

File: tests/codegen/params_count.rs
Comment (tgross35) [inline tests/codegen/params_count.rs]: Does this work? I don't think there is any need to assert the actual pattern of the variable names

```suggestion
// CHECK: (int32_t {{ \w* }}, int32_t {{ \w* }}, int32_t {{ \w* }})
```

File: tests/codegen/params_count.rs
Comment (tgross35) [inline tests/codegen/params_count.rs:12]: I think `CHECK-LABEL` should only be used when there are other `CHECK`s associated with it, otherwise a normal `CHECK` is fine

```suggestion
// CHECK: int32_t foo
// CHECK: int32_t main
```

File: tests/codegen/ret_value.rs
Comment (tgross35) [inline tests/codegen/ret_value.rs]:
```suggestion
// CHECK: return 42;
```

File: tests/bless/basic_math.c
Comment (tgross35) [inline tests/bless/basic_math.c:20]: Why is `int32_t` generated rather than `int`? On some (uncommon) platforms, `int` is 16 bits.

Comment (Wybxc) [reply tests/bless/basic_math.c:20]: This originates from Rust's `i32`.

 In Rust, there is no exact equivalent to C's `int` type (`core::ffi::c_int` is as an alias for either `i32` or `i16`, depending on the platform, and it is resolved before code generation).

Comment (tgross35) [reply tests/bless/basic_math.c:20]: Ah makes sense, thanks for the clarification.

File: tests/codegen/params_count.rs
Comment (Wybxc) [reply tests/codegen/params_count.rs:12]: `CHECK-LABEL` is required here to skip the function declaration. Otherwise, the `CHECK-LABEL` at line 13 would match the function declaration instead of the function definition, resulting in an error.

File: tests/codegen/params_count.rs
Comment (Wybxc) [reply tests/codegen/params_count.rs]: Neither `\w*` nor `[\w\d_]*` works for me.

(However, [regex101](https://regex101.com/r/sdmi5U/1) accepts this form. Maybe a bug in FileCheck?)

Comment (tgross35) [reply tests/codegen/params_count.rs]: Ah yeah, filecheck uses [ERE](https://en.wikibooks.org/wiki/Regular_Expressions/POSIX-Extended_Regular_Expressions). I think something like `[[:alnum:]_]*` should work but if not, just leave it as `[a-zA-Z0-9_]*` (I think it's okay to not check that the first character isn't a number)

File: bootstrap/src/test.rs
Comment (tgross35) [reply bootstrap/src/test.rs]: I see you updated this, looks great

**Review by tgross35 — APPROVED** [review]
Nothing else stands out to me so I think we should merge this, then follow up with issues for any problems that come up. Please resolve the last remaining open discussion then clean up history.

**Pull request merged**
